File: adapters/OathAndCoin.Content/Scenarios/ContrastRunner.cs
using System.Collections.Immutable;
using System.Text.Json.Nodes;
using OathAndCoin.Simulation;
using OathAndCoin.Simulation.Commands;
using OathAndCoin.Simulation.Decisions;
using OathAndCoin.Simulation.Ids;
using OathAndCoin.Simulation.State;

namespace OathAndCoin.Content.Scenarios;

/// <summary>
/// What running a <see cref="ContrastDefinition"/> found: the action each
/// branch reached, whether that pair matches the direction the definition
/// declared, and the RNG ordinal each branch actually spent its one decision
/// on.
/// </summary>
/// <param name="Flipped">
/// <c>true</c> only when the two actions match <see cref="ContrastDefinition.Expect"/>'s
/// own direction exactly — not merely "the two answers differ". A contrast
/// declaring <c>decline_to_accept</c> whose branches actually went
/// <c>accept</c>/<c>decline</c> (the opposite way) is not flipped as
/// declared, even though its two answers disagree; treating any disagreement
/// as success would let a contrast pass while asserting the wrong story
/// about its own content.
/// </param>
/// <param name="ActionFrom">What the hero decided in the <c>from</c> branch.</param>
/// <param name="ActionTo">What the hero decided in the <c>to</c> branch.</param>
/// <param name="OrdinalUsedFrom">
/// <see cref="GameMetadata.NextDecisionOrdinal"/> <em>after</em> the
/// <c>from</c> branch's one command — that is, how many RNG ordinals the
/// branch actually spent reaching its answer.
/// </param>
/// <param name="OrdinalUsedTo">The same, for the <c>to</c> branch.</param>
/// <remarks>
/// Read after the command, not before. Both branches start from a freshly
/// built state at ordinal 0, so reading it beforehand compared 0 with 0 and
/// would have stayed equal no matter what either branch went on to do —
/// including the one divergence this field exists to catch, where one branch
/// takes the gate path (spends nothing) and the other scores and draws a
/// mood (spends one). Read afterwards, "the same ordinal on both sides"
/// (HERO_DECISION_SPEC §7.3) is a claim about the two runs rather than about their shared
/// starting point: it says the difference between the branches is the varied
/// input, not a different roll of the same dice.
/// </remarks>
public sealed record ContrastResult(
    bool Flipped, ContentId ActionFrom, ContentId ActionTo, ulong OrdinalUsedFrom, ulong OrdinalUsedTo);

/// <summary>
/// Runs a <see cref="ContrastDefinition"/>'s two branches and reports what
/// each one decided.
/// </summary>
/// <remarks>
/// Both branches are built from the same <see cref="ContentSet"/> instance,
/// the same seed, and the same freshly-created <see cref="GameState"/> — the
/// only difference between them is the single field
/// <see cref="ContrastDefinition.Input"/> names, changed from
/// <see cref="ContrastDefinition.From"/> to <see cref="ContrastDefinition.To"/>.
/// Each branch then issues exactly one <see cref="ProposeContractToHero"/>
/// command, to the same hero, for the same contract, at command id 1 — no
/// other command runs in either branch. That is what makes "same seed, same
/// ordinal, same pre-state" a fact about how this method is built rather
/// than an author's promise: there is no code path here that could apply a
/// second command, use a different seed per branch, or reuse state mutated
/// by anything other than the one named field.
/// </remarks>
public static class ContrastRunner
{
    public static ContrastResult Run(ContrastDefinition definition)
    {
        ArgumentNullException.ThrowIfNull(definition);

        var content = ContentSet.Load(definition.ContentRoot);

        var from = RunBranch(content, definition, definition.From);
        var to = RunBranch(content, definition, definition.To);

        var flipped = definition.Expect switch
        {
            "decline_to_accept" => from.Action == Actions.Decline && to.Action == Actions.Accept,
            "accept_to_decline" => from.Action == Actions.Accept && to.Action == Actions.Decline,
            _ => throw new InvalidOperationException(
                $"Contrast '{definition.Name}' declares expect '{definition.Expect}', which "
                + "ContrastDefinition.Build should already have rejected."),
        };

        return new ContrastResult(flipped, from.Action, to.Action, from.Ordinal, to.Ordinal);
    }

    /// <summary>
    /// Builds one branch's initial state, applies <paramref name="value"/> to
    /// the field <see cref="ContrastDefinition.Input"/> names, and asks the
    /// definition's hero the definition's one question.
    /// </summary>
    private static (ContentId Action, ulong Ordinal) RunBranch(
        ContentSet content, ContrastDefinition definition, JsonNode value)
    {
        var state = content.CreateInitialState(definition.Seed, ScenarioRunner.RulesetVersion);
        state = ApplyVary(state, definition.Input, definition.Contract, value);

        var heroId = ResolveHeroId(state, definition.Hero, definition.Name);

        var engine = new SimulationEngine();
        var result = engine.Apply(
            state,
            new ProposeContractToHero(
                CommandId: 1,
                heroId,
                definition.Contract,
                ExpectedStateVersion: state.Metadata.StateVersion));

        if (!result.Applied || result.Decision is null)
        {
            throw new InvalidOperationException(
                $"Contrast '{definition.Name}' could not ask hero '{definition.Hero}' about contract "
                + $"'{definition.Contract}': the one command this branch issues was rejected "
                + $"('{result.RejectionCode}'). A contrast's single command must always be answerable, or "
                + "the two branches are not actually asking the same question.");
        }

        return (result.Decision.SelectedAction, result.State.Metadata.NextDecisionOrdinal);
    }

    /// <summary>
    /// Applies <paramref name="value"/> to exactly the field
    /// <paramref name="input"/> names on <paramref name="contractId"/>'s
    /// state, leaving every other field — including every other contract and
    /// every hero — untouched. This is the one place a vary actually changes
    /// anything; everything else about the two branches comes from the same
    /// <see cref="ContentSet.CreateInitialState"/> call.
    /// </summary>
    private static GameState ApplyVary(GameState state, string input, ContentId contractId, JsonNode value)
    {
        if (!state.Contracts.TryGetValue(contractId, out var contract))
        {
            throw new InvalidDataException(
                $"Content root has no contract definition '{contractId}' for this contrast to vary.");
        }

        var varied = input switch
        {
            "contract.payment" => contract with { Payment = ContrastDefinition.AsInt(value) },
            "contract.risk" => contract with { Risk = ContrastDefinition.AsInt(value) },
            "contract.tags" => contract with
            {
                Tags = ImmutableSortedSet.CreateRange(ContrastDefinition.AsContentIds(value)),
            },
            "contract.accepted_by" => WithAcceptedBy(state, contract, ContrastDefinition.AsContentIds(value)),
            _ => throw new InvalidOperationException(
                $"Unknown vary input '{input}', which ContrastDefinition.Build should already have rejected."),
        };

        return state with { Contracts = state.Contracts.SetItem(contractId, varied) };
    }

    /// <summary>
    /// The one input that varies a precondition rather than a content value
    /// (see the remarks on <see cref="ContrastDefinition"/>): who has already
    /// accepted this contract before the named hero is asked. Set directly on
    /// the freshly-built state rather than through a real
    /// <see cref="ProposeContractToHero"/> command for each accepting
    /// comrade — issuing one would be a second command in this branch, which
    /// is exactly the thing that would put the "no other command" guarantee
    /// in the hands of whoever writes the next contrast instead of in this
    /// method's own structure.
    /// </summary>
    /// <remarks>
    /// <see cref="ContractState.Status"/> is recomputed by the same rule
    /// <see cref="SimulationEngine.Apply"/> uses when a hero's own acceptance
    /// fills the crew — reaching <see cref="ContractState.RequiredCrew"/>
    /// closes the offer. Skipping this would let a precondition claim a
    /// contract is still <see cref="ContractStatus.Offered"/> when the real
    /// game would already have marked it <see cref="ContractStatus.Crewed"/>,
    /// and the named hero's question would be one the game could never
    /// actually reach. A crew this full also means
    /// <see cref="SimulationEngine.Apply"/> rejects the branch's own command
    /// with <see cref="RejectionCodes.ContractAlreadyResolved"/> — which
    /// <c>RunBranch</c> already turns into a loud
    /// <see cref="InvalidOperationException"/> rather than a silent, wrong
    /// answer.
    /// </remarks>
    private static ContractState WithAcceptedBy(
        GameState state, ContractState contract, ImmutableArray<ContentId> acceptedByDefinitions)
    {
        var heroIds = ImmutableSortedSet.CreateRange(
            acceptedByDefinitions.Select(definitionId => ResolveHeroId(state, definitionId, contract.Id.Value)));

        return contract with
        {
            AcceptedBy = heroIds,
            RespondedBy = contract.RespondedBy.Union(heroIds),
            Status = heroIds.Count >= contract.RequiredCrew ? ContractStatus.Crewed : contract.Status,
        };
    }

    /// <summary>
    /// Finds the runtime <see cref="HeroId"/> <see cref="ContentSet.CreateInitialState"/>
    /// assigned to <paramref name="heroDefinition"/>, rather than recomputing
    /// its content-id-sorted index independently — a second statement of that
    /// ordering rule here would drift from the loader's the day either one
    /// changes without the other.
    /// </summary>
    private static HeroId ResolveHeroId(GameState state, ContentId heroDefinition, string context)
    {
        foreach (var (id, hero) in state.Heroes)
        {
            if (hero.Definition == heroDefinition)
            {
                return id;
            }
        }

        throw new InvalidDataException(
            $"Content root has no hero definition '{heroDefinition}' for '{context}' to resolve.");
    }
}

File: packages/content/src/scenarios/contrast-runner.ts
import {
  Actions,
  SortedMap,
  SortedSet,
  compareContentIds,
  compareHeroIds,
  contractOf,
  createContractState,
  decide,
  type ContentId,
  type ContractState,
  type GameState,
  type HeldTrait,
  type HeroId,
  type HeroState
} from '@oath-and-coin/simulation';

import type { ContentSet } from '../content-set.ts';
import { createInitialState } from '../initial-state.ts';

import { ContrastExpectation, type ContrastDefinition, type ContrastVary } from './contrast-definition.ts';
import { RULESET_VERSION } from './scenario-runner.ts';

/**
 * The contrast runner (`HERO_DECISION_SPEC` §7.3, `MVP_PLAN` §5.5): builds two initial
 * states from one content set and one seed, differing only in the input a
 * {@link ContrastDefinition} names, and asks the same question — the same hero, about the
 * same contract, on the same ordinal — of both. Seed, ordinal and every field but the named
 * one match **by construction**, so a flipped answer cannot be explained by mood or by a
 * different pre-state (`ContrastRunner_UsesTheSameSeedAndOrdinalOnBothSides`).
 *
 * `decide` is called directly rather than through `proposeContractToHero`: the command
 * refuses anyone but the offer's key hero while a package is still a draft
 * (`RejectionCodes.NotTheKeyHero`, `NEGOTIATION_SPEC` §3.1), and a contrast has no
 * negotiation to stage first — it is one question, asked once, on a package built to hold
 * exactly the one difference it names.
 */

/** One side's answer, with what it was asked under. */
export interface ContrastSideResult {
  readonly seed: bigint;
  readonly ordinal: bigint;
  readonly action: ContentId;
}

/** Both sides of a contrast, and whether the answer flipped the way it was declared to. */
export interface ContrastRun {
  readonly left: ContrastSideResult;
  readonly right: ContrastSideResult;
  /**
   * Whether the answer moved from `left` to `right` in exactly the direction
   * {@link ContrastDefinition.expect} declares — never merely "the answer differed".
   * `accept → decline` on a contrast declaring `decline_to_accept` is not the declared
   * flip and this is `false` for it, the same as it is for no flip at all
   * (`ContrastRunner_DoesNotCountAMismatchedDirectionAsFlipped`).
   */
  readonly flipped: boolean;
}

/**
 * The baseline negotiation package every contrast composes before applying its one named
 * difference (`NEGOTIATION_SPEC` §2.1, §4). Not the engine's own starting offer
 * ({@link import('../initial-state.ts').createInitialState}'s `initialOffer()`, which pays
 * nothing and asks nobody) — a contrast is one question asked of one hero, and an unasked
 * question is not what any of the eight allowed inputs describe changing.
 *
 * - `advance` defaults to the contract's own `patronFee` — the guild's plain,
 *   unrenegotiated offer pays the whole published fee — unless `offer.advance` is the
 *   input under test, in which case the vary's own value is used directly. This is also
 *   what makes a `contract.patron_fee` contrast observable at all: `decide` never reads
 *   `patronFee` itself (`NEGOTIATION_SPEC` §4 — "The patron fee itself never appears"),
 *   only `offer.advance`, so varying the fee has to move what "the whole fee" means for
 *   the swing to exist.
 * - `promisedBonus` defaults to `0` — no promise on the table — unless `offer.promised_bonus`
 *   is the input under test (the vary's value), or `hero.believes_guild_promises` is (the
 *   contract's `patronFee`, so a contrast about belief has an actual promise to revoke;
 *   varying belief around a promise of zero would vary nothing `decide` reads).
 * - `methodTag` defaults to `null` — no method negotiated — unless `offer.method_tag` is
 *   the input under test.
 * - `keyHero` is always the hero being asked. Harmless where no promise is on the table
 *   (`decide`'s `trustedBonus` only reads `keyHero` at all when `promisedBonus > 0`), and
 *   required wherever one is: `createContractState` refuses `promisedBonus > 0` with
 *   `keyHero === null`.
 * - `grievance` and `believesGuildPromises` default to the campaign's own starting values
 *   (`0`, `true` — `initial-state.ts`) unless the corresponding `hero.*` input is under
 *   test.
 */
function buildSide(
  baseContract: ContractState,
  baseHero: HeroState,
  vary: ContrastVary,
  side: 'from' | 'to'
): { readonly contract: ContractState; readonly hero: HeroState } {
  let patronFee = baseContract.patronFee;
  let risk = baseContract.risk;
  let tags = baseContract.tags;
  let advanceOverride: number | null = null;
  let methodTag: ContentId | null = null;
  let promisedBonusOverride: number | null = null;
  let grievance = baseHero.grievance;
  let believesGuildPromises = baseHero.believesGuildPromises;

  switch (vary.input) {
    case 'contract.patron_fee':
      patronFee = vary[side];
      break;
    case 'contract.risk':
      risk = vary[side];
      break;
    case 'contract.tags':
      tags = SortedSet.from(compareContentIds, vary[side]);
      break;
    case 'offer.advance':
      advanceOverride = vary[side];
      break;
    case 'offer.method_tag':
      methodTag = vary[side];
      break;
    case 'offer.promised_bonus':
      promisedBonusOverride = vary[side];
      break;
    case 'hero.grievance':
      grievance = vary[side];
      break;
    case 'hero.believes_guild_promises':
      believesGuildPromises = vary[side];
      break;
  }

  const advance = advanceOverride ?? patronFee;
  const promisedBonus =
    promisedBonusOverride ?? (vary.input === 'hero.believes_guild_promises' ? patronFee : 0);

  const contract = createContractState({
    ...baseContract,
    patronFee,
    risk,
    tags,
    offer: {
      ...baseContract.offer,
      keyHero: baseHero.id,
      advance,
      methodTag,
      promisedBonus,
      respondedBy: SortedSet.empty(compareHeroIds),
      acceptedBy: SortedSet.empty(compareHeroIds)
    }
  });

  const hero: HeroState = { ...baseHero, grievance, believesGuildPromises };

  return { contract, hero };
}

/** The hero's traits, resolved through the campaign's trait rulebook, sorted by id. */
function resolveTraits(state: GameState, hero: HeroState): readonly HeldTrait[] {
  const traitIds = SortedSet.from(compareContentIds, hero.traits);

  return traitIds.values().map((traitId) => {
    const trait = state.traitRules.get(traitId);
    if (trait === undefined) {
      throw new Error(
        `Hero '${hero.definition}' carries trait id '${traitId}', but the campaign's traitRules has ` +
          'no entry for it — a content-loading bug, not a hero with no opinion.'
      );
    }

    return trait;
  });
}

/** The runtime id of the hero a contrast names, resolved by content id. */
function heroRuntimeId(state: GameState, definition: ContrastDefinition): HeroId {
  for (const [id, hero] of state.heroes.entries()) {
    if (hero.definition === definition.hero) {
      return id;
    }
  }

  throw new Error(
    `Contrast '${definition.contrast}' names hero '${definition.hero}', which the content at ` +
      `'${definition.contentRoot}' does not define.`
  );
}

function sideResult(
  state: GameState,
  heroState: HeroState,
  baseContract: ContractState,
  traits: readonly HeldTrait[],
  vary: ContrastVary,
  side: 'from' | 'to'
): ContrastSideResult {
  const { contract, hero } = buildSide(baseContract, heroState, vary, side);

  const decision = decide({
    hero,
    contract,
    traits,
    // A freshly-built contrast package has never been offered, so nobody has accepted it
    // yet — `acceptedBy` is always empty (see `buildSide`) and there is no comrade for the
    // rule's bonds term to find. A contrast about a hero's bonds is a different hero on
    // the same contract, not a mutated `acceptedBy` (`HERO_DECISION_SPEC` §7.3).
    crew: SortedMap.empty(compareHeroIds),
    campaignSeed: state.metadata.campaignSeed,
    decisionOrdinal: state.metadata.nextDecisionOrdinal,
    traceId: state.metadata.nextTraceId
  });

  return {
    seed: state.metadata.campaignSeed,
    ordinal: state.metadata.nextDecisionOrdinal,
    action: decision.result.selectedAction
  };
}

function isDeclaredFlip(
  left: ContentId,
  right: ContentId,
  expect: ContrastExpectation
): boolean {
  if (expect === ContrastExpectation.DeclineToAccept) {
    return left === Actions.Decline && right === Actions.Accept;
  }

  return left === Actions.Accept && right === Actions.Decline;
}

/**
 * Runs one contrast against `content`: builds the campaign's starting state at the
 * contrast's own seed, then asks the named hero the same question twice, differing only in
 * the one input the contrast declares.
 *
 * @throws if the contrast names a hero or contract `content` does not define, or if
 * building either side's package violates a `ContractState` invariant (`offer-state.ts`) —
 * for instance, a `contract.patron_fee` this low leaves an `offer.advance` this high outside
 * `0..patronFee`.
 */
export function runContrast(definition: ContrastDefinition, content: ContentSet): ContrastRun {
  const state = createInitialState(content, definition.seed, RULESET_VERSION);

  const heroId = heroRuntimeId(state, definition);
  const heroState = state.heroes.get(heroId);
  if (heroState === undefined) {
    throw new Error(`Contrast '${definition.contrast}' resolved hero '${definition.hero}' to no state.`);
  }

  const baseContract = contractOf(state, definition.contract);
  const traits = resolveTraits(state, heroState);

  const left = sideResult(state, heroState, baseContract, traits, definition.vary, 'from');
  const right = sideResult(state, heroState, baseContract, traits, definition.vary, 'to');

  return {
    left,
    right,
    flipped: isDeclaredFlip(left.action, right.action, definition.expect)
  };
}
